python/cairo-addons/src/cairo_addons/testing/runner.py

```diff
@@ -245,24 +266,39 @@ def _run(entrypoint, *args, **kwargs):
             else:
                 arg_value = kwargs[arg_name] if arg_name in kwargs else args[i]
                 stack.append(gen_arg(python_type, arg_value))
+
+        # ============================================================================
+        # STEP 4: SET UP EXECUTION CONTEXT AND LOAD MEMORY
+        # - Rationale: Finalize the stack with return pointers, set initial VM registers,
+        #   and load program/data into memory to start execution.
+        # - Add the dummy last fp and pc to the public memory, so that the verifier can enforce
+        #   [fp - 2] = fp.
+        # ============================================================================
         return_fp = runner.execution_base + 2
-        # Return to the jmp rel 0 instruction added previously
-        end = runner.program_base + len(runner.program.data) - 2
-        # Proof mode expects the program to start with __start__ and call main
-        # Adding [return_fp, end] before and after the stack makes this work both in proof mode and normal mode
+        end = runner.program_base + len(runner.program.data) - 2  # Points to jmp rel 0
         stack = [return_fp, end] + stack + [return_fp, end]
```

Comment:
not sure if we want to keep this, which is only to maintain compatibility with the `proof_mode` at compile time.